gui: Make device ID read-only and hide "nearby devices" when adding a new pending device (fixes #8083)

tomasz1986 opened the pull request:
Currently, there is no distinction between adding "new" and "pending"
devices. For this reason, the user is always presented with a list of
"nearby devices" to choose from. This commit adds such distinction to
the code, and in the case of "pending" device, both the device ID is
made read-only and the nearby devices list is hidden.

As a by-product of the function rename and clean-up, this commit also
hides the non-functional "remove" button that is shown when editing
device defaults.

Signed-off-by: Tomasz Wilczyński <twilczynski@naver.com>

### Screenshots

#### Before

![image](https://user-images.githubusercontent.com/5626656/149946382-cafb89c0-9d7e-4186-a191-d7849816d65f.png)

#### After

![image](https://user-images.githubusercontent.com/5626656/149946281-c297288c-200e-4287-906d-f3c5ef0d22dc.png)

acolomb commented:
Haven't tried myself, but won't this allow to click on another listed device ID which will replace the one supposed to be added? With this change there is not even an easy way to get it back.

I think it would be better to hide the whole suggestion list and make the ID input read-only when invoked from the "new device wants to connect" notification.

tomasz1986 commented:
> Haven't tried myself, but won't this allow to click on another listed device ID which will replace the one supposed to be added? With this change there is not even an easy way to get it back.

If replaced (or modified so that it doesn't match the other one anymore), the ID goes back to the list of nearby devices automatically.

acolomb commented:
Well I don't see anything wrong with keeping it in the list. Note also that the list is hard limited to five entries, so whether a specific device ID ends up there is pretty random when more local devices are in reach. I'd be disappointed when clicking on a few discovered device IDs to find at some point the one I really wanted did not come back to the list. Maybe not a real-world problem, but on the other hand I just don't see what we gain by hiding it.

The other thing (locking ID entry and hiding the list when called from a notification), that would be a worthwhile change IMHO.

tomasz1986 commented:
> Note also that the list is hard limited to five entries, so whether a specific device ID ends up there is pretty random when more local devices are in reach. I'd be disappointed when clicking on a few discovered device IDs to find at some point the one I really wanted did not come back to the list.

I didn't know about the limitation, but the list seems to be sorted alphabetically. It is that five random devices are picked first, and then sorted, or rather that everything is sorted first, and only then the five top ones are displayed? If it is the former, then yes, this change may be problematic. If it is the latter though, it shouldn't make any difference, should it?

> on the other hand I just don't see what we gain by hiding it.

The main problem for me here is not technical, but language-related really. The user is asked to "also select one of these nearby devices" even if the exact same ID is already filled in above. I'd argue that either the same ID below should be hidden (as this commit does), or alternatively the language needs to be changed (but doing so will also force to re-translate everything). As it is right now, I personally find the sentence confusing.

> The other thing (locking ID entry and hiding the list when called from a notification), that would be a worthwhile change IMHO.

Yeah, this would be nice, but I think this could be a material for a different commit 🙂.

imsodin commented:
I agree with @acolomb: I don't like hiding UI elements, and it doesn't seem particularly wrong to me if the device ID stays in the list after I clicked it - the explanation text applied before I clicked it. If anything the selected ID could be marked/disabled somehow, but that seems overly complicated for very little gain. Hiding this section entirely from the start when adding a specific device from the "invitation banner" would be good though.

tomasz1986 commented:
> If anything the selected ID could be marked/disabled somehow, but that seems overly complicated for very little gain.

This is already done thanks to the anchor underline staying once clicked, so I don't think any additional marking is needed here.

> Hiding this section entirely from the start when adding a specific device from the "invitation banner" would be good though.

I've modified the PR to do that, and also to make the ID read-only. I've also removed the `datalist` element (explained in the commit message) and fixed the invalid HTML (since a `p` can't contain block elements).

acolomb commented:
> I haven't hidden the "The device ID to enter here…" explanation under the ID. Should I do it too or just leave it be?

IMO yes, definitely.  The device ID was filled in automatically and cannot be changed.  So giving advice where to find it for entering manually is moot.

tomasz1986 commented:
> IMO yes, definitely. The device ID was filled in automatically and cannot be changed. So giving advice where to find it for entering manually is moot.

Done. I've also tried to adapt the PR to the newly added default ignore patterns. I've also removed the duplicated code related to editing existing devices and merged it with the one used here, as it seems to be the same case - fixed, read-only device ID and "nearby devices" hidden.

acolomb commented on gui/default/syncthing/device/editDeviceModalView.html:
```diff
@@ -40,14 +37,6 @@
               </p>
             </div>
-            <div ng-if="editingDeviceExisting()" class="input-group">
-              <div class="well well-sm text-monospace form-control" style="height: auto;" select-on-click>{{currentDevice.deviceID}}</div>
```
I see no need to change this, the existing devices can still use the "Well" element, while adding uses the regular input (which gets set to read-only when adding from the notification).

tomasz1986 replied:
Is there any purpose not to use a read-only `input` element though? As far as I see it, it makes no difference regarding functionality and design (and semantically `input` is the appropriate element anyway).

acolomb replied:
I remember now, it had to do with wrapping mainly.  The well word-wraps the ID nicely on narrow screens, while the input just truncates it.  That's an important distinction when looking at an existing device and trying to find out its ID.  But again, when adding a suggested ID, the user only wants it to be the one from the notification.  The need to see every character of it vanishes.

Having said that, why not just use the original code for read-only (the .well thing) also in case it's added from a notification?  Seems appropriate because you need none of the other stuff, namely the discovery list.

tomasz1986 replied:
The .well thing has been reused now, but the code has also been moved to the main section instead of having a separate one at the bottom. I'm marking this as "resolved" as the same conversation is already going on above, so I think it's better to keep everything in one place.

acolomb reviewed (changes requested):
You made some overly intrusive (and some probably unintentional) changes along the way.  Try to keep your logic changes inside the GUI minimal for each step.  Reviewing is already quite complicated in the HTML code.  I'd recommend ignoring whitespace diff (add `?w=1` to the [URL on Github](https://github.com/syncthing/syncthing/pull/8091/files?w=1)), then you'll notice some mistakes.

tomasz1986 commented on gui/default/syncthing/device/editDeviceModalView.html:
```diff
@@ -10,19 +10,16 @@
+                  <span class="fas fa-qrcode"></span>&nbsp;<span translate>Show QR</span>
+                </button>
               </div>
-              <datalist id="discovery-list">
```
Yes, I explained this above (and in the commit message).

Does this element serve any purpose other than to list the exact same set of device IDs as already listed below under the "nearby devices"?

acolomb replied:
Sorry I missed the commit message.  A real pity that GitHub is so focused on PR conversations that most of commit messages are ususally not shown.  I use them quite extensively as well to document code changes.

Back to the matter: The datalist helps when navigating with the keyboard for example.  Start typing some characters inside the field, then rely on "auto-complete" behaviour.  At least there is no reason to remove the feature.

tomasz1986 replied:
> Back to the matter: The datalist helps when navigating with the keyboard for example. Start typing some characters inside the field, then rely on "auto-complete" behaviour. At least there is no reason to remove the feature.

To be frank, as it is right now, I personally find the datalist very annoying. It pops up as soon as you click on the device id input field, and while doing so, it also covers the actually useful list of nearby devices in the process.

![image](https://user-images.githubusercontent.com/5626656/149639606-4bea80e3-363c-4b14-bc16-dd7630e7454b.png)

The device IDs have their last part cut-off too, which makes it look kind of broken. It does help with keyboard autocomplete (although, at the same time, it's also very easy to access the nearby devices list through the TAB key), but how about reducing the listed entries to just one, so that autocomplete still works, but the list doesn't hover anything useful below?

acolomb replied:
Note that this could look completely different depending on which browser.  Mobile devices or screen readers may make a totally different use of the data provided.  The app just says: "This may be meaningful inputs for the field", how the browser uses or renders it is up to them.

> To be frank, as it is right now, I personally find the datalist very annoying. It pops up as soon as you click on the device ID input field, and while doing so, it also covers the actually useful list of nearby devices in the process.

When someone explicitly clicks into the text entry field, I think they already decided to use the keyboard and not choose one from the list.  But the list may help to notice the string they're entering is actually among the discovered entries.  Again, think about a smartphone context.  The whole GUI might be invisible now because the virtual keyboard popped up to take 2/3 of the screen.

You may not need the feature, but there is no strong reason to remove it IMHO.

tomasz1986 replied:
> When someone explicitly clicks into the text entry field, I think they already decided to use the keyboard and not choose one from the list

Not really, I find myself often clicking on one of the nearby devices first, but then deciding not to use it and trying to remove the input ID manually. Doing so immediately triggers the datalist to pop up (which then results in the annoying covering of the nearby devices).

> Again, think about a smartphone context. The whole GUI might be invisible now because the virtual keyboard popped up to take 2/3 of the screen.

The situation on a smartphone is actually worse. This is a screenshot from an actual Android phone.

![image](https://user-images.githubusercontent.com/5626656/149640553-6ac37b66-0b5b-49de-b65d-0605fcd3be81.png)

The list is cut off even more, and the position is messed up too. Of course, there will be differences depending on the OS and the browser, but as it stands right now, I'd say it gets in the way much more than it actually helps.

It's also impossible to improve the situation by forcing the text to wrap or something, because browsers apparently still don't allow any styling of `datalist` and `option`. The current display is totally dependent on how each browser wishes to do it, with no control over it in the Web developer's hands.

I have just tested modifying the datalist to show only a single entry, but it still ended up covering the first ID in the nearby devices, so it's not really useful, at least with the current positioning.

acolomb replied:
What's wrong with covering the nearby devices list when you decided not to need one of those IDs?

How (badly) browsers render it should not mean we don't hand out the information to e.g. screen readers.  I don't think trying to style the popup or limiting it to one entry (which one?) to be a sensible solution.

tomasz1986 replied:
> What's wrong with covering the nearby devices list when you decided not to need one of those IDs?

The problem is I haven't. I'm still undecided, but I've clicked on the input field first (maybe out of habit). Then I thought that maybe the device I want to add has already been listed by Syncthing automatically. However, now the nearby devices list is already covered by the other one.

Yes, I know I could then rely on the datalist instead to choose the device in question, but I personally find the whole thing confusing, because at this point there are basically two lists displayed on the screen, none of which is fully visible. The nearby devices list shows only the last parts of the device IDs, while the datalist, displayed on top of everything else, shows the device IDs cut-off.

Basically, with the two lists displayed at once, there is no place, where we can see the full device IDs.

> How (badly) browsers render it should not mean we don't hand out the information to e.g. screen readers.

While true, I'd personally be very surprised if the current UI was usable by an actual blind person with a screen reader. There is too much invalid HTML paired up with JS-generated content, and Bootstrap and Angular aren't known for being accessibility-friendly either.

Nevertheless, I wouldn't have any problem with the datalist if it didn't cover the other useful content (and, to a lesser extent, at least be able to display the device IDs in full).

acolomb replied:
> However, now the nearby devices list is already covered by the other one.

Hit escape when you want to get rid of some value selection popup?  My point is that what covers what is out of our control here.  Giving the data to the browser to hopefully do something useful is the best we can do.

> Basically, with the two lists displayed at once, there is no place, where we can see the full device IDs.

If the one you want is actually one of the nearby devices, you only need the first (or last) few characters to determine which one.  Or the identicon added recently.  A truncated device ID is very much good enough for a list of max. 5 discovered devices.

tomasz1986 replied:
I don't think we're going to reach an agreement here. I still find the two lists displayed one on top of the other confusing. In the current state, it's difficult to focus on either of them. Yes, I know it's possible to press ESC or click somewhere on the screen to get rid of the pop-up. That's what I've actually been doing all the time so far. It still doesn't change the fact that the whole behaviour is very annoying (for me).

I may still remove this part of the commit just to make the whole PR more specific (and be more in line with the title). If there's no consensus regarding the datalist, I'm probably going to stick to getting rid of it in my personal builds only.

tomasz1986 replied:
I'm marking this one "resolved", as I've got rid of the datalist related change, but I'd also like to add that not only are the nearby devices covered by the pop-up, but also the warnings and the explanation are covered too.

Just have a look at the screenshots below.

Warnings covered:

![image](https://user-images.githubusercontent.com/5626656/149661243-68868dc0-23d3-46bb-b0f7-726517d59d90.png)
![image](https://user-images.githubusercontent.com/5626656/149661253-52098771-98a7-450b-abdf-6e061666bc82.png)

Explanation covered:

![image](https://user-images.githubusercontent.com/5626656/149661257-1e4acf17-6911-4330-98db-c29d426665bc.png)

tomasz1986 commented on gui/default/syncthing/device/editDeviceModalView.html:
```diff
@@ -10,19 +10,16 @@
         <div id="device-general" class="tab-pane in active">
           <div ng-if="!editingDeviceDefaults()" class="form-group" ng-class="{'has-error': deviceEditor.deviceID.$invalid && deviceEditor.deviceID.$dirty}" ng-init="loadFormIntoScope(deviceEditor)">
             <label translate for="deviceID">Device ID</label>
-            <div ng-if="!editingDeviceExisting()">
```
It's not the same, but isn't using an `input` tag (with a readonly attribute) here more appropriate than a `div`? That's the background of this change. This still is an input field after all (even if non-editable).

acolomb replied:
I dunno, didn't design it in the first place.  There were some discussions about it in #7653 and in the end we decided to keep it as is.  The select-on-click behavior is the main difference I suppose.

acolomb replied:
See the other comment, I now think it's better to switch to the original "well" version in your specific case.

tomasz1986 replied:
I have switched to the "well" version now, but I've also grouped everything together instead of having two separate sections. This way it's possible to reuse the same "Show QR" button. Please let me know what you think of it.

tomasz1986 commented on gui/default/syncthing/core/syncthingController.js:
```diff
@@ -1679,7 +1679,11 @@ angular.module('syncthing.core')
                         $scope.currentDevice.deviceID = deviceID;
-                        $scope.currentDevice._editing = "add";
+                        if (deviceID) {
+                            $scope.currentDevice._editing = "pending";
```
> When adding another value to this "kind of enum" variable, you must make sure that every other use of it will not change semantics with the new value. Did you check that?

I think so. This variable is used regarding adding new devices (both empty and pending), editing existing devices, and also editing default devices. Everything seems to be working as intended, but if you can see something wrong or a flaw in the code, please let me know.

> And I think this function should be refactored to completely avoid retrieving the discovery list at all for the case when deviceID is already provided.

Yeah, but I doubt I've got enough JS skills to do it. I did try to modify the code to move the logic further above at first, but it didn't really work as I wanted it to, so in the end I just decided to add small modifications to the existing code instead.

One problem I encountered was that if the discovery part was skipped completely in the case of pending devices, and then we added a new device from scratch (i.e. not a pending one) first, then the list would be remembered and then still present when adding a new pending device later. This means that the code still had to clear out the list on every pending folder adding. All in all, the code got messy very quickly and I wasn't really capable of fixing it cleanly.

acolomb replied:
We're all here to learn... :wink:  You might get some inspiration from the recent changes regarding default ignores, where retrieving the setting can be delayed if the user wants to edit them in the following dialog.  In code:
https://github.com/syncthing/syncthing/blob/main/gui/default/syncthing/core/syncthingController.js#L1965

You basically have a variable with a Promise that immediately resolves initially.  If the circumstances require, you replace it with a promise that depends on the HTTP query.  The following code just calls `.then()` on the variable, which works in both cases to delay executing the given code until the promise is resolved.  Pretty elegant.  I don't see any reason why the same should not work here, your troubles were probably caused by getting confused with the Promise API and what gets executed when.

> I think so. This variable is used regarding adding new devices (both empty and pending), editing existing devices, and also editing default devices. Everything seems to be working as intended, but if you can see something wrong or a flaw in the code, please let me know.

What I meant was: Please do check every single line of the *code* where `currentDevice._editing` is accessed.  I will review that part myself as well once the logical structure is finalized.

tomasz1986 replied:
I've checked all the lines. Everything seems fine for me (which doesn't mean it's 100% flaw-free, but I've done as much as I'm able to). I've got to say though that I'm not going to attempt any major refactoring of the code in this commit (so that the discovery list isn't needlessly retrieved). If you or anyone else wants to work on it, please proceed (probably in a separate PR, I'd assume).

I don't think such a refactoring is exactly related to this specific PR, because the same can be said for the already existing code that builds the discovery list even when editing existing or default devices, where it's unnecessary already.

imsodin replied:
This code is definitely terrible, mixing `success` and `then`, and returning a premise, but not the one called in the `then`. That's not on this PR though indeed, I can clean it up and skip the discovery call in case of adding a pending device in a followup unless you want to dig into it.

acolomb replied:
I agree, it was a thing with room for improvement before this PR and it can be fixed after it.

tomasz1986 commented on gui/default/syncthing/device/editDeviceModalView.html:
```diff
@@ -10,19 +10,16 @@
-              </datalist>
-              <p class="help-block" ng-if="discovery && discovery.length !== 0">
+            </div>
+            <div ng-if="currentDevice._editing === 'add'">
```
> And please use equal (`==`) checking instead of identity (`===`) for strings.

Fixed.

I haven't done anything with the internal variables though. Is it absolutely necessary? The `currentDevice._editing` is already present in the currently existing code (see https://github.com/tomasz1986/syncthing/blob/8d5f68c857fe0d85a0db349b52222ea8b3ffbd30/gui/default/syncthing/device/editDeviceModalView.html#L168), so I've just tried to follow a similar convention here.

imsodin commented on gui/default/syncthing/device/editDeviceModalView.html:
```diff
@@ -10,19 +10,20 @@
+                  <span class="fas fa-qrcode"></span>&nbsp;<span translate>Show QR</span>
+                </button>
               </div>
               <datalist id="discovery-list">
                 <option ng-repeat="id in discovery" value="{{id}}" />
               </datalist>
```
This needs a condition to only show when adding a new device as far as I see.

tomasz1986 replied:
Sure, it's fixed now. For the record, the element wasn't actually shown on the screen without its corresponding `input`.

imsodin commented on gui/default/syncthing/device/editDeviceModalView.html:
```diff
@@ -10,19 +10,20 @@
+                  <span class="fas fa-qrcode"></span>&nbsp;<span translate>Show QR</span>
+                </button>
               </div>
               <datalist id="discovery-list">
                 <option ng-repeat="id in discovery" value="{{id}}" />
               </datalist>
```
This should go into the conditional `div` below to only be set when adding a device.

tomasz1986 commented:
> This code is definitely terrible, mixing `success` and `then`, and returning a premise, but not the one called in the `then`. That's not on this PR though indeed, I can clean it up and skip the discovery call in case of adding a pending device in a followup unless you want to dig into it.

I'd be glad if you cleaned everything up in the followup PR 🙂.

imsodin reviewed (dismissed):
lgtm - handing to @acolomb to update if his review was addressed to satisfaction ;)

acolomb reviewed (changes requested):
This is starting to look really good I think.  Thanks for your continued effort despite some discussion... :wink:  Still some copy-paste errors in the details unfortunately.

tomasz1986 commented on gui/default/syncthing/device/editDeviceModalView.html:
```diff
@@ -10,19 +10,20 @@
-                </div>
+            <div class="input-group">
+              <input ng-if="currentDevice._editing == 'add'" name="deviceID" id="deviceID" class="form-control text-monospace" type="text" ng-model="currentDevice.deviceID" required="" valid-deviceid list="discovery-list" aria-required="true" />
+              <div ng-if="currentDevice._editing != 'add'" name="deviceID" id="deviceID" class="well well-sm form-control text-monospace" ng-model="currentDevice.deviceID" valid-deviceid style="height: auto;" select-on-click>{{currentDevice.deviceID}}</div>
```
Thanks for all the checking. I want to say this clearly though. There is no blind copy & paste here.

These attributes, i.e. `name`, `id`, `ng-model`, and `ng-model` are needed in order for the QR code validation to work. This is how the button is reused in the current iteration of the code. Without the attributes, the ID within the `div` is always "invalid", and as a result, the QR code button is also always greyed out in those instances.

acolomb replied:
Yeah, but if it's read-only anyway, there is no point in checking validity.  The user has no control over it, and we won't ever get an invalid device ID from a pending or existing device.

tomasz1986 replied:
That's true, but is there a better way that would get rid of the validation and still be able to use a single QR code button for everything?

I'd really like not to have duplicated code related to the button, especially since I'm also working on the other PR that adds a "copy to clipboard" and "send by email" buttons. I think the code would be way easier to maintain if only a single instance of the buttons existed (rather than having one with validation, and another one without it for every button).

acolomb replied:
As mentioned below, you need to adjust the `ng-disabled` attribute.

    ng-disabled="currentDevice._editing == 'add' && !deviceEditor.deviceID.$valid"

Or using the suggested wrapper function:

    ng-disabled="editingDeviceAdditional() && !deviceEditor.deviceID.$valid"

tomasz1986 replied:
This makes sense. Thank you very much 🙂. I'll try to implement all the fixes and then come back, probably sometime in the next few days.

tomasz1986 commented on gui/default/syncthing/device/editDeviceModalView.html:
```diff
@@ -40,14 +41,6 @@
-            <div ng-if="editingDeviceExisting()" class="input-group">
-              <div class="well well-sm text-monospace form-control" style="height: auto;" select-on-click>{{currentDevice.deviceID}}</div>
-              <div class="input-group-btn">
-                <button type="button" class="btn btn-default" data-toggle="modal" data-target="#idqr">
```
> As it is now, "Show QR" is always disabled when editing an existing device.

This isn't correct, is it? I've rechecked this right now. The button is enabled both when editing existing devices, and also when adding a pending device. It's disabled only when adding a new device from scratch and the ID is either empty or invalid.

acolomb replied:
Sorry, that was under the assumption that the extra attributes were removed, which I did before testing.

tomasz1986 commented on gui/default/syncthing/core/validDeviceidDirective.js:
```diff
@@ -4,7 +4,7 @@ angular.module('syncthing.core')
             link: function (scope, elm, attrs, ctrl) {
                 ctrl.$parsers.unshift(function (viewValue) {
-                    if (scope.currentDevice._editing != "add") {
+                    if (scope.currentDevice._editing.indexOf('add') != 0) {
```
I've removed the check altogether. What do you think of the current state?

acolomb replied:
Looks fine in this aspect.

tomasz1986 commented on gui/default/syncthing/device/editDeviceModalView.html:
```diff
@@ -172,7 +165,7 @@
       <span class="fas fa-times"></span>&nbsp;<span translate>Close</span>
     </button>
-    <div ng-if="has(['existing', 'defaults'], currentDevice._editing)" class="pull-left">
+    <div ng-if="editingDeviceExisting()" class="pull-left">
```
I just want to add a quick comment here. This is explained in the commit message too, as a part of the removing `currentDevice._editing` from the HTML code, I've also cleaned up the code here. The reason for removing "defaults" is and keeping only "existing" is that right now, there is a "non-functional" remove button when editing device defaults. Clicking it does nothing apart from closing the modal. I don't think this is intentional, and as I see it, the only case where the button makes sense is when editing existing devices.

acolomb replied:
Regarding "stuff explained in the commit message": I'd love to have more of that and I usually write very detailed messages myself.  But in this project it's kind of usual practice to just jot down "wip" or something because it will get squashed and discarded anyway.  That trains people to not look at the individual commits at all I suppose.  Unlike e.g. Linux development, where a PR usually concerns a "patch series".

tomasz1986 replied:
Yeah, although I always copy and paste the content of the main commit message into the PR description too. I also try to keep the message itself as descriptive and detailed as possible. One problem with this approach is that in order to actually edit the commit message I've got to rebase the whole commit history, and then do a force-push instead of just adding newer commit on top of the existing ones.

acolomb replied:
The commit message on your branch does not really matter.  PRs are squash-merged here, so the message for the actual commit in official Syncthing history is written by whoever merges it.  GitHub concatenates all individual commit messages only as a draft to start from (e.g. remove all "wip" lines).

Rebasing helps to keep that draft clean, but in general it's usually more confusing during review.  I often checkout PRs locally for testing, which works with added commits (fast-forward), but not after rebasing.  So best avoided unless you have a good reason (e.g. several logical changes within one PR that each deserve a detailed commit message).

tomasz1986 replied:
From my observation, it depends who merges the commit 😉. The "wips" are always removed, but the commit message itself is often left with no changes. For example, see https://github.com/syncthing/syncthing/commit/0fb7cc186c0ba7262a6cd1404d6aa1c812b7f529 and https://github.com/syncthing/syncthing/commit/abdf02451705a4b606f2991bec2ea631f34229fe. In the first case, both the commit message and the PR description matched, and the commit message was left in the final squash and merge. In the second case though, I updated only the PR description and forgot to update the commit message, which was then still used in the final squash and merge commit, despite not 100% matching the final iteration of the PR.

This is one of the reasons why I prefer to keep both the main commit message and the PR description have the exact same text, and also update both when needed (which is only possible through a `rebase` and `force-push`).

acolomb replied:
If it was my project, I'd have each PR submitter push a cleaned-up history, which then gets merged without squashing and always with a merge commit.  More like the "patch series" discipline on the Linux kernel for example.  But that's not this team's policy and not a discussion to be had here :wink:

acolomb commented on gui/default/syncthing/core/syncthingController.js:
```diff
@@ -1610,6 +1610,10 @@ angular.module('syncthing.core')
         }
 
+        $scope.editingDeviceNew = function() {
+            return $scope.currentDevice._editing == 'new';
```
A comment stating that "new-pending" was intentionally left out here would be good.

tomasz1986 replied:
I've now added a helper function for `pending` too. It's not used at the moment though. I'm not sure if it's better to still have it or to replace it with a comment instead.

acolomb replied:
Well, adding a pending device **is** adding a new device.  Could you please just apply the comment as given?

* `new-pending` instead of `pending`.
* No extra function that is never called.
* Comment in `editingDeviceNew()` that it intentionally consideres only `new`, but not `new-pending`.

tomasz1986 replied:
Done.

acolomb replied:
The comment was not supposed to be about a "missing" function. Suggestion:

    // The "new-pending" value is intentionally disregarded here.

Please add **within** the function, as I wrote in my last comment.

tomasz1986 replied:
Done.

acolomb commented on gui/default/syncthing/core/syncthingController.js:
```diff
@@ -1679,7 +1683,9 @@ angular.module('syncthing.core')
                         $scope.currentDevice = p.data;
                         $scope.currentDevice.name = name;
                         $scope.currentDevice.deviceID = deviceID;
-                        $scope.currentDevice._editing = "add";
+                        if (!deviceID) {
+                            $scope.currentDevice._editing = "new";
+                        };
```
You're not setting the variable at all now, while previously you had the special `add-pending` value.  That's error prone, please go back to setting it explicitly.

I don't see the need to rename the value from "add" to "new".  It doesn't really matter what the value is, but it's usually good to keep the changes minimal.  Especially if you want to get the review process done in a reasonable time frame :wink:

tomasz1986 replied:
The variable is now set. I've kept the `new` name though. The reason behind it is that the helper functions can be named `editingDeviceNew` and `editingDevicePending`, which seem, at least for me, easier to understand than something like `editingDeviceAdd` and `editingDeviceAddPending` or similar.

What do you think?

acolomb reviewed (changes requested):
Looking even better and it worked as intended during my (short) testing.  Fewer lines of code is good.  One thing you should not have removed from the PR, see comment below.

acolomb commented:
Do we already have an associated issue that could be linked for the release notes?

tomasz1986 commented:
> Do we already have an associated issue that could be linked for the release notes?

The issue #8083 is linked to in the title.
